### PR TITLE
[SPARK-27545] [SQL] Uncache table needs to delete the temporary view …

### PR DESCRIPTION
**Pull request by httfighter**
…created when the cache table is executed

## What changes were proposed in this pull request?

When executing the uncache table, delete the temporary table created when the cache table is executed.

## How was this patch tested?

 unit tests

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 1]
```diff
@@ -32,6 +31,8 @@ import org.apache.spark.sql.execution.datasources.{CreateTable, DataSource}
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.storage.StorageLevel
 
+import scala.util.Try
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Hi, @httfighter . Could you run the Scala style check first?
```
$ dev/scalastyle
```

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 2]
```diff
@@ -442,7 +443,11 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
   override def uncacheTable(tableName: String): Unit = {
     val tableIdent = sparkSession.sessionState.sqlParser.parseTableIdentifier(tableName)
     val cascade = !sessionCatalog.isTemporaryTable(tableIdent)
-    sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableName), cascade)
+    val wasCached = Try(sparkSession.catalog.isCached(tableIdent.unquotedString)).getOrElse(false)
+    if (wasCached) {
+      sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableIdent), cascade)
+      sparkSession.sessionState.catalog.dropTempView(tableIdent.table)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: I'm not sure this is the correct fix. Could you add a UT, please?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 1 above)
Comment (httfighter) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: @dongjoon-hyun Thank you! I have fixed it.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 2 above)
Comment (httfighter) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: @dongjoon-hyun OK!

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Thanks!

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 3]
```diff
@@ -32,7 +33,6 @@ import org.apache.spark.sql.execution.datasources.{CreateTable, DataSource}
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.storage.StorageLevel
 
-
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: We usually avoid this kind of unrelated change.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 4]
```diff
@@ -442,7 +442,11 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
   override def uncacheTable(tableName: String): Unit = {
     val tableIdent = sparkSession.sessionState.sqlParser.parseTableIdentifier(tableName)
     val cascade = !sessionCatalog.isTemporaryTable(tableIdent)
-    sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableName), cascade)
+    val wasCached = Try(sparkSession.catalog.isCached(tableIdent.unquotedString)).getOrElse(false)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Why use Try? Will `isCached` throw exception?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: Btw, `isCached` is also in `CatalogImpl`. Why to access it via `sparkSession.catalog`?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 5]
```diff
@@ -442,7 +442,11 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
...
+    if (wasCached) {
+      sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableIdent), cascade)
+      sparkSession.sessionState.catalog.dropTempView(tableIdent.table)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: If the table is not a temporary view, no need to drop it.

**Review by viirya — COMMENTED** [review]
I think we can improve the title and description of this PR. When we run `cache table as select`, it actually creates a temp view. Now when uncaching it, the temporary view still exists.

Semantically, dropping the temp view when uncaching it sounds correct, because it is created by the cache command. But this is behavior change, we need to think more about it.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala [hunk 6]
```diff
@@ -442,7 +442,11 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
   override def uncacheTable(tableName: String): Unit = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: I don't think we should put dropping view at `uncacheTable`. `cacheTable` in the same `CatalogImpl` doesn't create view.

`CacheTableCommand` creates view for `cache table as select` query. So I think it is more correct to drop the view in `UncacheTableCommand`.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 4 above)
Comment (httfighter) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: @viirya Thank you! 
I didn't notice  the `isCached `method in `CatalogImpl`. Therefore, I mimicked the usage in `tables.scala` to avoid exceptions.
  In` tables.scala`, 
```
      // If an exception is thrown here we can just assume the table is uncached;
      // this can happen with Hive tables when the underlying catalog is in-memory.
      val wasCached = Try(sparkSession.catalog.isCached(oldName.unquotedString)).getOrElse(false)
```
Now I am modifying to use the `isCached  `provided by `CatalogImpl ` without considering this exception.

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 3 above)
Comment (httfighter) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: @viirya Ok!

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 5 above)
Comment (httfighter) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: @viirya `dropTempView`  will only delete the temporary view. Are you suggesting that I increase a judgment on the tempory view here?

File: sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala (hunk 6 above)
Comment (httfighter) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: I have also considered this issue. Finally, I decided to it in `uncacheTable` for two reasons.
1. `UncacheTableCommand` only uses `sparkSession.catalog.uncacheTable` to do the operation of deleting the cache, and there is no other operation.
2.  
There are many other places that use `sparkSession.catalog.uncacheTable` to implement the operation of `uncache table`.  I think these places also need to consider drop the view.

@viirya May I ask if these considerations are correct?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: If the view isn't created when caching, I don't think it is good to remove the view when uncaching. `CacheTableCommand` is the caching command creating the view. I think it makes more sense to remove the view at `UncacheTableCommand`. They are a pair of commands.

When you do deleting view at `uncacheTable`. For a view which isn't created by `cache table as select`, uncaching it will delete it too? This sounds not correct to me.

Comment (httfighter) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala]: @viirya Yes. There is indeed this problem. However, there is still a problem of accidental deletion when removing the view at `UncacheTableCommand`. Because when executing `uncache table`, I can't judge whether the temporary view is created by the user or created when the `cache table` is executed.

I want to ask, why create a view when executing `cache table as select` . What does this view do? Because there is no correspondence between `view `and `plan`.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
Gentling ping @httfighter . Please don't forget the valid test case (https://github.com/apache/spark/pull/24440/files#r277545786).

**Comment by HyukjinKwon** [comment]
Gentling ping @httfighter

**Comment by httfighter** [comment]
@HyukjinKwon @dongjoon-hyun There is still a problem I don't know how to deal with it.
As he said @viirya ，When you do deleting view at uncacheTable. For a view which isn't created by cache table as select, uncaching it will delete it too.

**Comment by dongjoon-hyun** [comment]
@httfighter . Given the situation, shall we close this as 'Won't Do' because the user can do `uncache table` and `drop temp view` manually?

**Comment by httfighter** [comment]
@dongjoon-hyun There will also be problems with accidental deletion. Need to distinguish between the view created by the cache, or the view created by the user.

**Comment by dongjoon-hyun** [comment]
Sorry, but without a test case for that, I don't get a clear idea what you mean. We need to talk via code.

**Comment by httfighter** [comment]
@dongjoon-hyun OK. I will give a test case for the current modification first. Thank you.

**Comment by dongjoon-hyun** [comment]
Gentle ping, @httfighter .

**Comment by httfighter** [comment]
@dongjoon-hyun Thank you for reminding.I have added a test case.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by HyukjinKwon** [comment]
ok to test
